testing/mock/_asserts.ts

```diff
@@ -0,0 +1,324 @@
+/** This module is browser compatible. */
+
+import {
+  assertEquals,
+  AssertionError,
+  assertIsError,
+  assertRejects,
+} from "../asserts.ts";
+import { Spy, SpyCall } from "./mock.ts";
+
+/** An error related to spying on a function or instance method. */
+export class MockError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = "MockError";
+  }
+}
+
+/**
+ * Asserts that a spy is called as much as expected and no more.
+ */
+export function assertSpyCalls<
+  Self,
+  Args extends unknown[],
+  Return,
+>(
+  spy: Spy<Self, Args, Return>,
+  expectedCalls: number,
+) {
+  try {
+    assertEquals(spy.calls.length, expectedCalls);
+  } catch (e) {
+    assertIsError(e);
+    let message = spy.calls.length < expectedCalls
+      ? "spy not called as much as expected:\n"
+      : "spy called more than expected:\n";
+    message += e.message.split("\n").slice(1).join("\n");
+    throw new AssertionError(message);
+  }
+}
+
+/** Call information recorded by a spy. */
+export interface ExpectedSpyCall<
+  // deno-lint-ignore no-explicit-any
+  Self = any,
+  // deno-lint-ignore no-explicit-any
+  Args extends unknown[] = any[],
+  // deno-lint-ignore no-explicit-any
+  Return = any,
+> {
+  /** Arguments passed to a function when called. */
+  args?: [...Args, ...unknown[]];
+  /** The instance that a method was called on. */
+  self?: Self;
+  /**
+   * The value that was returned by a function.
+   * If you expect a promise to reject, expect error instead.
+   */
+  returned?: Return;
+  error?: {
+    /** The class for the error that was thrown by a function. */
+    // deno-lint-ignore no-explicit-any
+    Class?: new (...args: any[]) => Error;
+    /** Part of the message for the error that was thrown by a function. */
+    msgIncludes?: string;
+  };
+}
+
+/**
+ * Asserts that a spy is called as expected.
+ * Returns the call.
```

Comment:
I think returning value from assertion is confusing and less readable

(Note: We had similar discussion in the past about this design https://github.com/denoland/deno_std/pull/1052#discussion_r674510375

Reply:
Would you be good with keeping the third argument optional so that it can still be used for just checking if the spyFunc was called enough times but just having this assertion function changed to not returning a value?

Reply:
I went ahead and made this change, the assert functions no longer return anything.

Reply:
Thanks!